tempesta_fw/http_parser.c

```diff
@@ -498,153 +489,51 @@ enum {
-	}
-
-	/* Note, that in case of testing @msg->conn may be unitialized */
-	if (msg->conn && (TFW_CONN_TYPE(msg->conn) & Conn_Srv)) {
-		TfwHttpResp *resp = (TfwHttpResp *)msg;
-		if ((resp->s_line.flags & TFW_STR_COMPLETE) &&
-		    !tfw_str_eolen(&resp->s_line)) {
-			tfw_str_set_eolen(&resp->s_line, eolen);
-		}
-	}
-
-	return TFW_PASS;
-}
 
 #define RGEN_EOL()							\
 __FSM_STATE(RGen_EoL) {							\
-	parser->_tmp.eol = 0;						\
-	/* Fall through */						\
+	parser->_eol = 0;						\
+	/* fall through */						\
 }									\
-__FSM_STATE(RGen__EoL) {						\
-	if (likely(IS_CR_OR_LF(c))) {					\
-		/*							\
-		 * We use special register to track line endings. New	\
-		 * characters are appended to the beginning while old	\
-		 * characters are shifted left. The lower 4 bits used	\
-		 * to track CR/LF characters.				\
-		 */							\
-		parser->_tmp.eol = (parser->_tmp.eol << 4) | (c & 0xf);	\
-		TFW_DBG3("parser: eol %08lx\n", parser->_tmp.eol);	\
-									\
-		/*							\
-		 * We have a number of valid CR/LF mixtures. Any other	\
-		 * mixtures must be blocked:				\
-		 *							\
-		 *  LF          - next header / empty-line (incomplete)	\
-		 *  CR LF       - next header / empty-line (incomplete)	\
-		 *  CR          - (incomplete)				\
-		 *  LF CR       - empty-line (incomplete)		\
-		 *  LF LF       - empty-line				\
-		 *  LF CR LF    - empty-line				\
-		 *  CR LF CR    - empty-line (incomplete)		\
-		 *  CR LF LF    - empty-line				\
-		 *  CR LF CR LF - empty-line				\
-		 */							\
-		switch (parser->_tmp.eol) {				\
-		case 0xa:						\
-		case 0xda:						\
-			if (__handle_newline(msg))			\
-				return TFW_BLOCK;			\
-			/* Fall through */				\
-		case 0xd:						\
-		case 0xad:						\
-		case 0xaa:						\
-		case 0xada:						\
-		case 0xdad:						\
-		case 0xdaa:						\
-		case 0xdada:						\
-			goto good_looking_eol;				\
-		}							\
-									\
+__FSM_STATE(RGen_EoLine) {						\
+	parser->_eol = (parser->_eol << 4) | c;				\
+	TFW_DBG3("parser: eol %08lx\n", parser->_eol);			\
+	if (parser->_eol == 0xd)					\
+		__FSM_MOVE_nofixup(RGen_EoLine);			\
+	/* Allow only LF and CRLF as a newline delimiters */		\
+	if (parser->_eol != 0xa && parser->_eol != 0xda)		\
 		return TFW_BLOCK;					\
-									\
-good_looking_eol:							\
-									\
-		/*							\
-		 * Set empty-line mark only if LFxx or CRLFxx was	\
-		 * catched and crlf wasn't completed yet.		\
-		 */							\
-		if ((parser->_tmp.eol & 0xf0) == 0xa0) {		\
-			if (!(msg->crlf.flags & TFW_STR_COMPLETE))	\
-				__msg_crlf_open(p);			\
+	/* The header may be unopened in case of parsing s_line */	\
+	if (!parser->hdr.ptr)						\
+		__FSM_MOVE_nofixup(RGen_Hdr);				\
+	tfw_str_set_eolen(&parser->hdr, 1 + !!(parser->_eol == 0xda));	\
+	/* Zero length means that we've got an empty-line */		\
+	if (unlikely(!parser->hdr.len)) {				\
+		if (!(msg->crlf.flags & TFW_STR_COMPLETE)) {		\
+			msg->crlf = parser->hdr;			\
+			msg->crlf.flags |= TFW_STR_COMPLETE;		\
```

Comment:
`eolen` was set on hdr at line 523 and in case of empty-line detected (CRLF) whole the header will be copied to `crlf`. So, the `crlf->skb`, `crlf->len` (0) and `crlf->ptr` will be copied from the hdr.